File: week1_basic_data_structures/4_stack_with_max/StackWithMax.java
import java.util.*;
import java.io.*;

public class StackWithMax {
    class FastScanner {
        StringTokenizer tok = new StringTokenizer("");
        BufferedReader in;

        FastScanner() {
            in = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() throws IOException {
            while (!tok.hasMoreElements())
                tok = new StringTokenizer(in.readLine());
            return tok.nextToken();
        }

        int nextInt() throws IOException {
            return Integer.parseInt(next());
        }
    }

    public class StackMaxInt {
        class Pair {
            public int elem;
            public int max;

            public Pair(int elem, int max) {
                this.elem = elem;
                this.max = max;
            }
        }

        public Stack<Pair> s = new Stack<Pair>();

        public boolean empty() {
            return s.empty();
        }

        public int max() {
            if (s.empty())
                return 0;
            return s.peek().max;
        }

        public void push(int i) {
            int max_below = max();
            s.push(new Pair(i, Math.max(max_below, i)));
        }

        public void pop() {
            s.pop();
        }
    }

    public void solve() throws IOException {
        FastScanner scanner = new FastScanner();
        int queries = scanner.nextInt();
        // Stack<Integer> stack = new Stack<Integer>();
        StackMaxInt stack = new StackMaxInt();
        for (int qi = 0; qi < queries; ++qi) {
            String operation = scanner.next();
            if ("push".equals(operation)) {
                int value = scanner.nextInt();
                stack.push(value);
            } else if ("pop".equals(operation)) {
                stack.pop();
            } else if ("max".equals(operation)) {
                // System.out.println(Collections.max(stack));
                System.out.println(stack.max());
            }
        }
    }

    static public void main(String[] args) throws IOException {
        new StackWithMax().solve();
    }
}

File: week1_basic_data_structures/4_stack_with_max/stack_with_max.py
# python3
import sys
from collections import namedtuple

E = namedtuple("E", ["i", "sup"])


class StackWithMax():
    def __init__(self):
        self.__stack = []

    def empty(self):
        return not self.__stack

    def push(self, i):
        if self.__stack:
            new_sup = max(self.__stack[-1].sup, i)
            self.__stack.append(E(i, new_sup))
        else:
            self.__stack.append(E(i, i))

    def pop(self):
        assert(len(self.__stack))
        self.__stack.pop()

    def max(self):
        assert(len(self.__stack))
        return self.__stack[-1].sup


if __name__ == '__main__':
    stack = StackWithMax()

    num_queries = int(sys.stdin.readline())
    for _ in range(num_queries):
        query = sys.stdin.readline().split()

        if query[0] == "push":
            stack.push(int(query[1]))
        elif query[0] == "pop":
            stack.pop()
        elif query[0] == "max":
            print(stack.max())
        else:
            assert(0)
